gulpfile.js

```diff
@@ -124,28 +172,197 @@ gulp.task("default", function (callback) {
+        // dummy command that succeeds if emulator is ready and fails otherwise
+        spawnCommand("xcrun", ["simctl", "getenv", "booted", "asdf"], (code) => {
+            if (!code) return onEmulatorInit("iOS");
+            else {
+                console.log("iOS emulator is not ready yet!");
+                return onFailure();
+            }
+        }, true);
+    }
+    // kills the Android emulator then starts it
+    function killThenStartAndroid() {
+        spawnCommand("adb", ["emu", "kill"], () => {
+            // emulator @emulator, which starts the android emulator, never returns, so we must check its success on another thread
+            spawnCommand("emulator", ["@emulator"], undefined, false, true);
+        
+            var emulatorReadyAttempts = 0;
+            function androidEmulatorReadyLooper() {
+                ++emulatorReadyAttempts;
+                if (emulatorReadyAttempts > emulatorMaxReadyAttempts)
+                {
+                    console.log("Android emulator is not ready after " + emulatorMaxReadyAttempts + " attempts, abort.");
+                    androidProcess.kill();
+                    return callback(1);
+                }
+                setTimeout(androidEmulatorReady.bind(undefined, androidEmulatorReadyLooper), emulatorReadyCheckDelay);
+            }
+            androidEmulatorReadyLooper();
+        }, true);
+    }
+    // kills the iOS emulator then starts it
+    function killThenStartIOS() {
+        spawnCommand("killall", ["\"" + iOSSimulatorProcessName + "\""], () => {
+            spawnCommand("xcrun", ["instruments", "-w", iOSEmulatorName], () => {
+                var emulatorReadyAttempts = 0;
+                function iOSEmulatorReadyLooper() {
+                    ++emulatorReadyAttempts;
+                    if (emulatorReadyAttempts > emulatorMaxReadyAttempts)
+                    {
+                        console.log("iOS emulator is not ready after " + emulatorMaxReadyAttempts + " attempts, abort.");
+                        return callback(1);
+                    }
+                    setTimeout(iOSEmulatorReady.bind(undefined, iOSEmulatorReadyLooper), emulatorReadyCheckDelay);
+                }
+                iOSEmulatorReady(iOSEmulatorReadyLooper);
+            });
+        }, true);
+    }
+    if (!restartIfRunning) {
+        if (android) {
+            androidEmulatorReady(() => {
+                killThenStartAndroid();
+            });
+        }
+        if (ios) {
+            iOSEmulatorReady(() => {
+                killThenStartIOS();
+            });
+        }
+    } else {
+        if (android) killThenStartAndroid();
+        if (ios) killThenStartIOS();
+    }
+    // This needs to be done so that the task will exit.
+    // The command that creates the Android emulator persists with the life of the emulator and hangs this process unless we force it to quit.
+    gulp.doneCallback = (err) => {
+        process.exit(err ? 1 : 0);
+    }
+}
```

Comment:
Seems like collapsing `startEmulators(callback, restartIfRunning, platforms) {... ,killThenStart(platform ...), emulatorReady(platform ...)` etc together would shorten the code here because they pretty much do the same thing with some small differences (command args, killing android process).